docs/running-calyx/fud2/rhai-api.md

```diff
@@ -0,0 +1,152 @@
+
+In the generated Ninja code, `shell` will create both a `rule` wrapping the shell command and a `build` command that invokes that rule. When a `defop` contains multiple `shell` commands, `fud2` automatically generates Ninja dependencies among the `build` command to ensure they run in order.
+
+### shell_deps
+
+```
+shell_deps(<string>, [<dep1>, <dep2>, ...], [<target1>, <target2>, ..])
+```
+When called in the body of an op, that op will run `<string>` as a shell command if it needs to generate `<target1>, <target2>, ...` from `<dep1>, <dep2>, ...`. It is an error to call `shell_deps` outside of the body of an op. Additionally, it is an error to call `shell_deps` in the body of an op in which `shell`  was called prior.
+
+Targets and deps are either strings, such as `"file1"`, or identifiers, such as if `c: calyx` existed in the signature of an op then `c` could be used as a target or dep.
+
+A call to `shell_deps` corresponds directly to a Ninja rule in the Ninja file generated by `fud2`.
+
+## Low Level Rhai
+
+A common routine for using low level Rhai is:
+1. Define states
+2. Define setup functions that register variables and rules
+3. Define an emitter function that produces commands using variables and rules created from (2)
+4. Define an op that uses setup functions and emitter to produce the target state from the start state
+
+### Defining states
+
+```
+state(<name>, [<ext1>, <ext2>, .. ])
+```
+Defines a state with the name `<name>` where files can have extensions `<ext1>`, `<ext2>`, etc. `<name>` and extensions are all strings.
+
+### Defining operations
+
+```
+op(<op name>, [<setup1>, <setup2>, ..], <input state>, <output state>, <emit function>)
+```
+
+Defines an op with name `<op name>` that generates `<output state>` from `<input state>`. `<op name>` is a string, and `<input state>` and `<output state>` are states. 
+
+The op uses the setup functions `<setup1>, <setup2>, ..` to create vars and rules, and the `<emit function>` to produce commands.
+
+A setup function is simply a function that takes in an emitter.
+
+The emit function would have the following declaration:
+```
+|e, input, output| { ... }
+```
+where `e` is an emitter, and `input` and `output` are strings containing the filenames of the input and output of the operation.
+
+### Emitter API
+
+#### Creating variables
+
+```
+var_(<variable name>, <filename>)
+```
+Defines a Ninja variable `<variable name>` which refers to the file `<filename>`. Both `<variable name>` and `<filename>` are strings.
+
+<br>
+
+```
+config_var(<variable name>, <configuration path>)
+```
+Defines a Ninja variable `<variable name>` using the value obtained from `<configuration path>` from the configuration file. Both `<variable name>` and `<configuration path>` are strings.
```

Comment:
If the config value is undefined, raise an error. (The config value is required, not optional.)